lib/elixir/lib/calendar.ex

```diff
@@ -1332,11 +1595,46 @@ defmodule NaiveDateTime do
 
   """
   @spec compare(Calendar.naive_datetime, Calendar.naive_datetime) :: :lt | :eq | :gt
-  def compare(naive_datetime1, naive_datetime2) do
-    case {to_tuple(naive_datetime1), to_tuple(naive_datetime2)} do
-      {first, second} when first > second -> :gt
-      {first, second} when first < second -> :lt
-      _ -> :eq
+  def compare(%{calendar: calendar1} = naive_datetime1, %{calendar: calendar2} = naive_datetime2) do
+    if Calendar.compatible_calendars?(calendar1, calendar2) do
+      case {to_rata_die(naive_datetime1), to_rata_die(naive_datetime2)} do
+        {first, second} when first > second -> :gt
+        {first, second} when first < second -> :lt
+        _ -> :eq
+      end
+    else
+      raise ArgumentError, """
+      cannot compare #{inspect naive_datetime1} with #{inspect naive_datetime2}:
+      This comparison would be ambiguous as their calendars have incompatible day rollover moments.
+      Specify an exact time of day (using `DateTime`s) to resolve this ambiguity.
+      """
+    end
+  end
+
+  def convert(%{calendar: calendar} = naive_datetime, calendar) do
+    {:ok, naive_datetime}
+  end
+  def convert(%{calendar: ndt_calendar} = naive_datetime, calendar) do
+    if Calendar.compatible_calendars?(ndt_calendar, calendar) do
+      result_naive_datetime =
+        naive_datetime
+        |> to_rata_die
+        |> from_rata_die(calendar)
+      {:ok, result_naive_datetime}
+    else
+      {:error, :incompatible_calendars}
+    end
+  end
+
+  @spec convert!(NaiveDateTime.t, Calendar.calendar) :: NaiveDateTime.t
```

Comment:
Wrong spec. Remember it should use `Calendar.naive_datetime` as per above and not `NaiveDateTime.t`.